src/uid2Esp.ts

```diff
@@ -0,0 +1,61 @@
+export class UID2EncryptedSignalProvider {
+  constructor() {
+    if (
+      typeof window.getAdvertisingToken === "function" ||
+      // if the SDK has been initialized before this script loaded and enabled esp, the token will be
+      // send automatically as it might missed initComplete event 
+      this.isUID2SDKIntegrated()
+    ) {
+      this.sendSignal();
+    }
+  }
+
+  public sendSignal = () => {
+    const uid2Handler = this.retrieveAdvertisingTokenHandler();
+
+    if (!uid2Handler) {
+      console.warn("Please implement `getAdvertisingToken`");
+      return;
+    }
+
+    if (uid2Handler()) {
+      window.googletag = window.googletag || {
+        cmd: [],
+      };
+      window.googletag.encryptedSignalProviders =
+        window.googletag.encryptedSignalProviders || [];
+      window.googletag.encryptedSignalProviders.push({
+        id: "uidapi.com",
+        collectorFunction: () => Promise.resolve(uid2Handler()),
+      });
+    }
+  };
+
+  private isUID2SDKIntegrated = (): boolean => {
+    return Boolean(
+      window.__uid2 &&
+      "espEnabled" in window.__uid2 &&
+      window.__uid2.espEnabled
+    );
+  };
+
+  private retrieveAdvertisingTokenHandler = (): Function | undefined => {
+    if (typeof window.getAdvertisingToken === "function") {
+      return window.getAdvertisingToken!;
+    }
+    if (this.isUID2SDKIntegrated() && 'getAdvertisingToken' in window.__uid2!) {
+      return window.__uid2!.getAdvertisingToken!.bind(window.__uid2);
+    }
+  };
+}
+
+declare global {
+  interface Window {
+    __uid2Esp: UID2EncryptedSignalProvider;
```

Comment:
General style question: `UID2EncryptedSignalProvider` or `Uid2EncryptedSignalProvider`?

Reply:
We have both `UID2xxxx` and `Uid2xxx` in SDK 3, I was not sure which name convention to follow.  I will change it to `Uid2EncryptedSignalProvider`

Reply:
Yeah, I've noticed that inconsistency too - I'd like to fix up as many as possible.